clusterloader2/pkg/measurement/common/wait_for_controlled_pods.go

```diff
@@ -178,13 +182,31 @@ func (w *waitForControlledPodsRunningMeasurement) gather(syncTimeout time.Durati
 	if !w.isRunning {
 		return fmt.Errorf("metric %s has not been started", w)
 	}
-	desiredCount, maxResourceVersion, err := w.getObjectCountAndMaxVersion()
+	objectKeys, maxResourceVersion, err := w.getObjectKeysAndMaxVersion()
 	if err != nil {
 		return err
 	}
 
+	// Wait until checkers for all objects are registered:
+	// - when object is created/updated, it's enough to wait for its resourceVersion to
+	//   be processed by our handler; thus we wait until all events up to maxResourceVersion
+	//   are processed before proceeding
+	// - when object is deleted, by definition it will not be returned by the LIST request,
+	//   thus resourceVersion of the deletion may be higher than the maxResourceVersion;
+	//   we solve that by waiting until list of currently existing objects (that we propagate
+	//   via our handler) is equal to the expected one;
+	//   NOTE: we're not resiliant to situations where an object will be created/deleted
+	//   after the LIST call happened. But given measurement and phases don't infer with
+	//   each other, it can't be clusterloader that deleted it. Thus we accept this limitation.
+	//   NOTE: we could try waiting for the informer state to be the same and use the
+	//   resourceVersion from there, but then existence of bookmarks and the fact that our
+	//   informer doesn't necessary follow all objects of a given type can break that.
+	//   See #1259 for more details.
+
 	cond := func() (bool, error) {
-		return w.opResourceVersion >= maxResourceVersion, nil
+		w.lock.Lock()
+		defer w.lock.Unlock()
+		return w.opResourceVersion >= maxResourceVersion && objectKeys.Equal(w.objectKeys), nil
```

Comment:
To confirm, we cannot wait just for `objectKeys.Equal(...)` because it can easily break for the "update" case, right?

Reply:
Right - the maxRV is needed for update case.